File: Dynamic Programming/MinimizingCoins.py
s=input()
n,x=map(int,s.split())
s=input()
lis=list(map(int,s.split()))
# lis.sort()
INF=10000000
dp=[INF for _ in range(x+1)]
dp[0]=0
for i in range(1,x+1):
    for j in range(n):
        if i-lis[j]>=0 and dp[i-lis[j]]!=INF:
            dp[i]=min(dp[i],dp[i-lis[j]]+1)
if dp[x]==INF:
    print(-1)
else: 
    print(dp[x])

File: Dynamic Programming/MinimizingCoins.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
    int n,x;
    cin>>n>>x;
    vector<int>coins(n);
    for(int i=0;i<n;i++)
    {
        cin>>coins[i];
    }
    int INF=10000000;
    vector<int>dp(x+1,INF);
    dp[0]=0;
    for(int i=1;i<=x;i++)
    {
        for(int j=0;j<n;j++)
        {
            if(i-coins[j]>=0)
            {
                dp[i]=min(dp[i],dp[i-coins[j]]+1);
            }
        }
    }
    cout<< ((dp[x]==INF) ? -1 : dp[x]);
    return 0;
}